### PR TITLE
Fixed `When_receiving_messages_already_dispatched` test

### PR DESCRIPTION
**Pull request by SzymonPobiega**
* Renamed to `When_a_duplicate_message_arrives`
 * Replaced counting messages in the handler with a behavior that runs before the outbox. The old code would always return `2` no matter how many actual messages were send over the wire

### CODE REVIEW

File: src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    using NServiceBus.AcceptanceTests.EndpointTemplates;
+    using NServiceBus.AcceptanceTests.ScenarioDescriptors;
+    using NServiceBus.Configuration.AdvanceExtensibility;
+    using NServiceBus.Pipeline;
+    using NUnit.Framework;
+
+    public class When_a_duplicate_message_arrives : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task It_does_not_dispatch_messages_aready_dispatched()
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: the test naming sounds off?

docs states:
>  If previously processed, any messages in the outbox are delivered for that message, but do not invoke message-processing logic again

File: src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+        public class DownstreamEndpoint : EndpointConfigurationBuilder
+        {
+            public DownstreamEndpoint()
+            {
+                EndpointSetup<DefaultServer>(b =>
+                {
+                    b.LimitMessageProcessingConcurrencyTo(1);
+                    b.GetSettings().Set("DisableOutboxTransportCheck", true);
+                    b.EnableOutbox();
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: wouldn't it make more sense to disable the outbox on the `DownStreamEndpoint` and verify that 3 messages arrived instead?

**Comment by timbussmann** [comment]
code changes + renamings are hard to review when it doesn't the renaming properly. It's very helpful to split those into two commits (for the next time).

File: src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs (hunk 2 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: :+1:

File: src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs (hunk 1 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: Hmm I think the doco is off. If the messages have been already dispatched, they should not be dispatched again if we detect a duplicate. We should just silently ignore it.

Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: are you sure https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageProcessingConnector.cs#L51 ?

Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: Right now we lean on the persisters to clear the dispatched messages for us

https://github.com/Particular/NServiceBus.RavenDB/issues/81

We did have plans to make this explicit from a core perspective. @SzymonPobiega did we raise an issue for this?

In short: I think @SzymonPobiega is correct but the code is very confusing due to the implicitness

Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: Rename to "Should_not_dispatch_messages_already_dispatched"? (note the typo as well)

Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: good to know, thanks for clarification @andreasohlund! I agree, we should definitely improve that API :+1:

Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: I am not sure yet how to improve that API. With the new "nullify the operations on dispatcihng" approach, it is super easy to just return an empty list of transport operations to the core (if they are dispatched).

Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Reliability/Outbox/When_a_duplicate_message_arrives.cs]: I'll rename the test as @andreasohlund suggested.

**Comment by timbussmann** [comment]
@SzymonPobiega can you raise a PR or an issue to fix the docs regarding the outbox behavior?

**Comment by SzymonPobiega** [comment]
@timbussmann here you are: https://github.com/Particular/docs.particular.net/pull/1286

**Pull request merged**